Hola **Julio**!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

**Tip**: Para mejorar la presentación de tus notebooks es recomendable dejar siempre en la primera celda un título y una breve introducción al proyecto para que sea claro desde el inicio lo que se realizará
</div>

In [1]:
#importar librerias
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [2]:
#leer el dataset
df=pd.read_csv('/datasets/users_behavior.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [4]:
df.sample(5)

,calls,minutes,messages,mb_used,is_ultra
301,62.0,404.81,78.0,16948.05,0
2203,95.0,680.96,44.0,15326.86,0
118,28.0,215.41,0.0,6045.15,0
1686,102.0,714.82,34.0,18341.56,0
1174,35.0,248.36,77.0,21368.52,0


<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Antes de pasar a los procesos de entrenamiento procura siempre revisar la calidad de los datos con los que trabajarás 
</div>

In [5]:
#división de datos para el entrenamiento y la validación
df_train, df_valid = train_test_split(df, random_state=54321, test_size =0.25)

<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Muy bien realizada la división de los datos para evaluar correctamente los modelos. Es importante tener dividido los datos ya que no se puede evaluar el modelo sobre los mismos datos que ya vió para su entrenamiento
</div>

In [6]:
#segmentacion de dataset
features_train=df_train.drop('is_ultra',axis=1)
features_train

,calls,minutes,messages,mb_used
103,74.0,540.81,17.0,20075.65
1695,62.0,473.14,4.0,14623.29
1542,66.0,430.02,20.0,12049.53
636,92.0,628.47,72.0,16839.52
1216,44.0,318.99,48.0,14817.03
...,...,...,...,...
2020,81.0,506.46,21.0,7510.09
96,29.0,267.55,29.0,16996.83
3098,46.0,360.78,12.0,12706.94
650,41.0,244.59,39.0,0.00


In [7]:
target_train=df_train['is_ultra']
target_train

103     0
1695    1
1542    0
636     0
1216    0
       ..
2020    1
96      0
3098    1
650     1
2641    0
Name: is_ultra, Length: 2410, dtype: int64

In [8]:
features_valid=df_valid.drop('is_ultra',axis=1)
features_valid

,calls,minutes,messages,mb_used
1727,104.0,739.28,55.0,21940.11
2885,67.0,469.58,63.0,7952.07
2730,74.0,549.14,20.0,17160.61
899,31.0,205.14,22.0,10250.58
2119,26.0,172.87,16.0,6431.26
...,...,...,...,...
2031,52.0,385.11,0.0,17617.43
484,0.0,0.00,191.0,32448.02
2689,71.0,491.14,0.0,28677.71
68,138.0,1009.11,64.0,27807.13


In [9]:
target_valid=df_valid['is_ultra']
target_valid

1727    1
2885    1
2730    0
899     0
2119    0
       ..
2031    0
484     1
2689    1
68      1
720     0
Name: is_ultra, Length: 804, dtype: int64

In [10]:
#verificar arbol de desicion
best_result_tree = 0
best_depth = 0
for depth in range(1,20): 
    model = DecisionTreeClassifier(max_depth=depth,random_state=12345)
    model.fit(features_train,target_train) 
    predictions_valid = model.predict(features_valid)
    result =accuracy_score(target_valid, predictions_valid)
    if result > best_result_tree:
        best_result_tree = result
        best_depth = depth

print('mejor profundidad:',best_depth,'mejor resultado:', best_result_tree) 

mejor profundidad: 8 mejor resultado: 0.7935323383084577


In [11]:
#verificar bosque aleatorio
best_result_forest = 0
best_depth = 0
best_est = 0
for est in range(10, 51, 10):
    for depth in range (1, 21):
        model_forest =  RandomForestClassifier(max_depth=depth,n_estimators=est, random_state=12345) # inicializa el constructor de modelos con los parámetros random_state=54321, n_estimators=est y max_depth=depth
        model_forest.fit(features_train,target_train)
        predictions_valid = model_forest.predict(features_valid)
        result =accuracy_score(target_valid, predictions_valid)
        if result>best_result_forest:
            best_result_forest=accuracy_score(target_valid, predictions_valid)
            best_depth=depth
            best_est=est
print('mejor profundidad:',best_depth,'mejor resultado:', best_result_forest,'mejor cantidad de estimadores:',best_est)        

mejor profundidad: 11 mejor resultado: 0.8134328358208955 mejor cantidad de estimadores: 30


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Muy bien planteado el código para probar con los diferentes modelos para conseguir un mejor performance!
</div>

In [12]:
model = LogisticRegression(random_state=12345)
model.fit(features_train,target_train)
predictions_valid = model.predict(features_valid)
result =accuracy_score(target_valid, predictions_valid)
print('resultado:',result)  

resultado: 0.7251243781094527


se puede observar que el mejor modelo fue el RandomForestClassifier con una profundidad de 11 teninedo un accuracy de .8134 lo cual es uno mas que aceptable

In [13]:

#revolver el target
target_shuffled = np.random.permutation(target_train)
model_forest.fit(features_train, target_shuffled)
predictions_valid = model_forest.predict(features_valid)
print("Accuracy con etiquetas barajadas:", accuracy_score(target_valid, predictions_valid))


Accuracy con etiquetas barajadas: 0.6703980099502488


al ver que al revolver el target train baja el Accuracy a .67 demuestra que el modelo si esta aprendiendo patrones reales

<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Buen trabajo con tu proyecto! Entrenaste los modelos correctamente obteniendo una métrica por encima de lo propuesto.
    
Como recomendaciones, sería explorar más los datos y la calidad de los mismos, así como redactar conclusiones más detalladas de lo observado en un EDA y cómo podría estar relacionándose con los resultados de tus modelos.
    
Saludos!
</div>